api/api.py

```diff
@@ -0,0 +1,188 @@
+from flask import Flask, request, stream_with_context, Response
+from flask_restful import inputs, reqparse, Resource, Api
+import json
+import jsonschema
+import io
+import logging
+from PIL import Image
+import pkg_resources
+import threading
+import time
+
+import mcu.automation
+
+static_folder = pkg_resources.resource_filename(__name__, "/swagger")
+
+app = Flask(__name__, static_url_path="/", static_folder=static_folder)
+api = Api(app)
+automation_events = []
+
+
+class Events:
+    """This used to be the 'Automation Server'."""
+
+    def __init__(self):
+        self.clients = []
+        self.events = []
+        self.condition = threading.Condition()
+
+    def add_client(self, client):
+        app.logger.debug(f"events: new client")
+        self.clients.append(client)
+
+    def remove_client(self, client):
+        app.logger.debug(f"events: client exited")
+        self.clients.remove(client)
+
+    def broadcast(self, event):
+        app.logger.debug(f"events: broadcasting {event} to ({len(self.clients)}) client(s)")
+        self.events.append(event)
+        for client in self.clients:
+            client.send_screen_event(event)
+        with self.condition:
+            self.condition.notify_all()
+
+events = Events()
+
+def create_app(screen_, seph_):
+    global screen, seph
+    screen, seph = screen_, seph_
+    return app
+
+class Automation(Resource):
+    def post(self):
+        document = request.get_data()
+
+        try:
+            document = document.decode("ascii")
+        except UnicodeDecodeError:
+            return "invalid encoding", 400
+
+        if document.startswith("file:"):
+            return "invalid document", 400
+
+        try:
+            rules = mcu.automation.Automation(document)
+        except json.decoder.JSONDecodeError:
+            return "invalid document", 400
+        except jsonschema.exceptions.ValidationError:
+            return "invalid document", 400
+
+        seph.automation = rules
+
+        return {}, 200
+
+class Button(Resource):
+    def __init__(self):
+        self.parser = reqparse.RequestParser()
+        self.parser.add_argument("action", choices=["press", "release"])
+        self.parser.add_argument("delay", type=float, default=0.1)
+        super(Button, self).__init__()
+
+    def post(self):
+        args = self.parser.parse_args()
+        button = request.base_url.split("/")[-1]
+        if button == "both":
+            seph.handle_button(1, True)
+            seph.handle_button(2, True)
+            time.sleep(args.delay)
+            seph.handle_button(1, False)
+            seph.handle_button(2, False)
```

Comment:
This feels like a `if action is None` / `... actions[action]` is missing from the case `button == "both"`.

Reply:
`action` is required now